src/dbnode/storage/index/postings_list_cache_lru.go

```diff
@@ -0,0 +1,237 @@
+// Copyright (c) 2019 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package index
+
+import (
+	"container/list"
+	"errors"
+
+	"github.com/m3db/m3/src/m3ninx/postings"
+
+	"github.com/pborman/uuid"
+)
+
+// PostingsListLRU implements a non-thread safe fixed size LRU cache of postings lists
+// that were resolved by running a given query against a particular segment. Normally
+// an key in the LRU would look like:
+//
+// type key struct {
+//    segmentUUID uuid.UUID
+//    pattern     string
+//    patternType PatternType
+// }
+//
+// However, some of the postings lists that we will store in the LRU have a fixed lifecycle
```

Comment:
hm not sure why you need all this protection here. If a segment is closed, how would we attempt a `Get()` on it?

Reply:
Yeah I reasoned through the entire lifecycle with @robskillington and I think the conclusion we came to was that this wasn't necessary at all, but we didn't like the idea of leaving objects in the cache that if access would cause a segmentation fault. Since there was a cheap way to prevent that from happening we decided to be overly cautious

Reply:
lol OK. +1 for the paranoia =)

Maybe also describe that motivation in the file comment. Currently reads like it's necessary as opposed to additional safety

Reply:
cool updated the comment